src/api/tryGetFileTemporaryUrl.js

```diff
@@ -1,19 +1,49 @@
 /* @flow strict-local */
 import type { Auth } from './apiTypes';
+import { isUrlOnRealm } from '../utils/url';
 import getFileTemporaryUrl from './messages/getFileTemporaryUrl';
 
 /**
- * Get the complete authless URL to a realm uploaded file, returns null if that
- * fails.
+ * Like getFileTemporaryUrl, but on error returns null instead of throwing.
  *
- * See aslo jsdoc on getFileTemporaryUrl.
- *
- * @param href URL path to the resource.
- * @param auth Current user's auth information.
+ * Validates `url` to give getFileTemporaryUrl the kind of input it expects
+ * and returns null if validation fails.
  */
-export default async (href: string, auth: Auth): Promise<string | null> => {
+export default async (url: URL, auth: Auth): Promise<URL | null> => {
+  if (!isUrlOnRealm(url, auth.realm)) {
+    return null;
+  }
```

Comment:
Can we add this check before the commit that starts using `parsedUrl.pathname`? I.e. this one:

    api.getFileTemporaryUrl: Stop requesting with double slash, as in api/v1//…
    
    CZO today seems happy to accept the double slash, returning the same
    thing it returns when we don't send the double slash. Still, better
    not to send it, since it's not what zulip.yaml asks for.
    
    Although we don't expect user-visible changes when connecting to
    current servers, I haven't marked this NFC because it changes the
    mobile app's behavior as witnessed by the server.

That'd be helpful for limiting how wild the difference between (as of that commit) `href` and `pathname` can be. If `href` is an absolute-URL string, or a scheme-relative URL string, then switching from `href` to `pathname` means much more than adding a leading slash -- it strips off the whole prefix that gives the origin.

Thinking about those cases makes that commit complicated -- so since we don't actually want to handle them, it'd be handy to eliminate them first.

Reply:
I've made this change, but I think it still leaves behind some of the complicatedness you described.

In particular, in that commit you quoted, switching from `href` to `pathname` can still mean stripping off the whole prefix that gives the origin—it's just that that case is narrowed to when `href` is an absolute-URL string _on the realm_.

I'm not sure if, in the wild, `href` is ever actually an absolute-URL string on the realm. The old code would have mishandled that case, right: we'd fail to get a temp URL, then open `href` in a browser. That'd be annoying: since it's an on-realm URL, auth is probably required, meaning the user probably has to authenticate to Zulip in that browser. But as bug symptoms go, that one could be much worse, so I wouldn't expect much signal from users that would help us confirm this is happening.

Reply:
> In particular, in that commit you quoted, switching from `href` to `pathname` can still mean stripping off the whole prefix that gives the origin—it's just that that case is narrowed to when `href` is an absolute-URL string _on the realm_.

True. But in those cases, the new behavior is the desired behavior: the URL string was meant as a reference to e.g. `https://chat.example/user_uploads/1234/image.png`, we had been mishandling it, now we handle it the way one would want to handle a reference to `https://chat.example/user_uploads/1234/image.png`.

Whereas when it's off-realm, the new behavior is *not* desired behavior. So then one has to work out whether it's any worse than the old also-undesired behavior, and in particular whether it opens up any vulnerability that wasnt in the old behavior.

Specifically: if the URL string looks like `https://evil.h4x0r.example/user_uploads/1234/image.png`, then we had been mishandling it in a way that caused nothing to happen (the server would have nothing to say about the URL path `/api/v1/https://evil.h4xor.example/…`), but now we'd be mishandling it in a different way (by treating it the same as the perfectly appropriate URL string `/user_uploads/1234/image.png`) which would cause us to get farther in the subsequent logic. We'd be in a situation of thinking we were looking at the message's actual link of `https://evil.h4x0r.example/user_uploads/1234/image.png` but in fact looking at an entirely different resource at `https://chat.example/user_uploads/1234/image.png`. That situation seems ripe for a security vulnerability. By ensuring the URL is on-realm before we fix the other issue, we avoid ever setting up that situation.